README.md
<a href="https://hapi.dev"><img src="https://raw.githubusercontent.com/hapijs/assets/master/images/family.png" width="180px" align="right" /></a>

# @hapi/boom

#### HTTP-friendly error objects.

**boom** is part of the **hapi** ecosystem and was designed to work seamlessly with the [hapi web framework](https://hapi.dev) and its other components (but works great on its own or with other frameworks). If you are using a different web framework and find this module useful, check out [hapi](https://hapi.dev) – they work even better together.

### Visit the [hapi.dev](https://hapi.dev) Developer Portal for tutorials, documentation, and support

## Useful resources

- [Documentation and API](https://hapi.dev/family/boom/)
- [Version status](https://hapi.dev/resources/status/#boom) (builds, dependencies, node versions, licenses, eol)
- [Project policies](https://hapi.dev/policies/)
- [Free and commercial support options](https://hapi.dev/support/)


CHANGELOG.md
Breaking changes are documented using GitHub issues, see [issues labeled "breaking changes"](https://github.com/hapijs/boom/issues?q=is%3Aissue+label%3A%22breaking+changes%22).

If you want changes of a specific minor or patch release, you can browse the [GitHub milestones](https://github.com/hapijs/boom/milestones?state=closed&direction=asc&sort=due_date).


test/index.ts
import * as Boom from '..';
import * as Lab from '@hapi/lab';

const { expect } = Lab.types;


class X {

    x: number;

    constructor(value: number) {

        this.x = value;
    }
};

const decorate = new X(1);


// new Boom.Boom()

expect.type<Boom.Boom>(new Boom.Boom());
expect.type<Error>(new Boom.Boom());
expect.type<Boom.Boom>(new Boom.Boom('error'));

expect.error(new Boom.Boom('error', { decorate }));     // No support for decoration on constructor


class CustomError extends Boom.Boom {}
expect.type<Boom.Boom>(new CustomError('Some error'));


const boom = new Boom.Boom('some error');
expect.type<Boom.Output>(boom.output);
expect.type<Boom.Payload>(boom.output.payload);


// boomify()

const error = new Error('Unexpected input');

expect.type<Boom.Boom>(Boom.boomify(error, { statusCode: 400 }));
expect.type<Boom.Boom>(Boom.boomify(error, { statusCode: 400, message: 'Unexpected Input', override: false }));
expect.type<number>(Boom.boomify(error, { decorate }).x);

expect.error(Boom.boomify(error, { statusCode: '400' }));
expect.error(Boom.boomify('error'));
expect.error(Boom.boomify(error, { statusCode: 400, message: true }));
expect.error(Boom.boomify(error, { statusCode: 400, override: 'false' }));
expect.error(Boom.boomify());
expect.error(Boom.boomify(error, { decorate }).y);


// isBoom

expect.type<boolean>(Boom.boomify(error).isBoom);

// isBoom()

expect.type<boolean>(Boom.isBoom(error));
expect.type<boolean>(Boom.isBoom(Boom.boomify(error)));

expect.error(Boom.isBoom('error'));
expect.error(Boom.isBoom({ foo: 'bar' }));
expect.error(Boom.isBoom({ error: true }));
expect.error(Boom.isBoom());


// 4xx Errors

// badRequest()

expect.type<Boom.Boom>(Boom.badRequest('invalid query', 'some data'));
expect.type<Boom.Boom>(Boom.badRequest('invalid query', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.badRequest('invalid query'));
expect.type<Boom.Boom>(Boom.badRequest());

expect.error(Boom.badRequest(400));
expect.error(Boom.badRequest({ foo: 'bar' }));


// unauthorized()

expect.type<Boom.Boom>(Boom.unauthorized('invalid password'));
expect.type<Boom.Boom>(Boom.unauthorized('invalid password', 'simple'));
expect.type<Boom.Boom>(Boom.unauthorized(null, 'Negotiate', 'VGhpcyBpcyBhIHRlc3QgdG9rZW4='));
expect.type<Boom.Boom>(Boom.unauthorized('invalid password', 'sample', { ttl: 0, cache: null, foo: 'bar' }));
expect.type<Boom.Boom>(Boom.unauthorized('invalid password', 'sample', { ttl: 0, cache: null, foo: 'bar' } as Boom.unauthorized.Attributes));
expect.type<Boom.Boom>(Boom.unauthorized());
expect.type<Boom.Boom>(Boom.unauthorized('basic', ['a', 'b', 'c']));
expect.type<Boom.Boom & Boom.unauthorized.MissingAuth>(Boom.unauthorized('', 'basic'));
expect.type<Boom.Boom & Boom.unauthorized.MissingAuth>(Boom.unauthorized(null, 'basic'));
expect.type<boolean>(Boom.unauthorized('', 'basic').isMissing);
expect.type<boolean>(Boom.unauthorized(null, 'basic').isMissing);

expect.error(Boom.unauthorized(401))
expect.error(Boom.unauthorized('invalid password', 500))
expect.error(Boom.unauthorized('invalid password', 'sample', 500))
expect.error(Boom.unauthorized('basic', ['a', 'b', 'c'], 'test'));
expect.error(Boom.unauthorized('message', 'basic').isMissing);


// paymentRequired()

expect.type<Boom.Boom>(Boom.paymentRequired('bandwidth used', 'some data'));
expect.type<Boom.Boom>(Boom.paymentRequired('bandwidth used', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.paymentRequired('bandwidth used'));
expect.type<Boom.Boom>(Boom.paymentRequired());

expect.error(Boom.paymentRequired(402));
expect.error(Boom.paymentRequired({ foo: 'bar' }));


// forbidden()

expect.type<Boom.Boom>(Boom.forbidden('try again some time', 'some data'));
expect.type<Boom.Boom>(Boom.forbidden('try again some time', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.forbidden('try again some time'));
expect.type<Boom.Boom>(Boom.forbidden());

expect.error(Boom.forbidden(403));
expect.error(Boom.forbidden({ foo: 'bar' }));


// notFound()

expect.type<Boom.Boom>(Boom.notFound('missing', 'some data'));
expect.type<Boom.Boom>(Boom.notFound('missing', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.notFound('missing'));
expect.type<Boom.Boom>(Boom.notFound());

expect.error(Boom.notFound(404));
expect.error(Boom.notFound({ foo: 'bar' }));


// methodNotAllowed()

expect.type<Boom.Boom>(Boom.methodNotAllowed('this method is not allowed', 'some data'));
expect.type<Boom.Boom>(Boom.methodNotAllowed('this method is not allowed', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.methodNotAllowed('this method is not allowed'));
expect.type<Boom.Boom>(Boom.methodNotAllowed());

expect.error(Boom.methodNotAllowed(405));
expect.error(Boom.methodNotAllowed({ foo: 'bar' }));


// notAcceptable()

expect.type<Boom.Boom>(Boom.notAcceptable('unacceptable', 'some data'));
expect.type<Boom.Boom>(Boom.notAcceptable('unacceptable', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.notAcceptable('unacceptable'));
expect.type<Boom.Boom>(Boom.notAcceptable());

expect.error(Boom.notAcceptable(406));
expect.error(Boom.notAcceptable({ foo: 'bar' }));


// proxyAuthRequired()

expect.type<Boom.Boom>(Boom.proxyAuthRequired('auth missing', 'some data'));
expect.type<Boom.Boom>(Boom.proxyAuthRequired('auth missing', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.proxyAuthRequired('auth missing'));
expect.type<Boom.Boom>(Boom.proxyAuthRequired());

expect.error(Boom.proxyAuthRequired(407));
expect.error(Boom.proxyAuthRequired({ foo: 'bar' }));


// clientTimeout()

expect.type<Boom.Boom>(Boom.clientTimeout('timed out', 'some data'));
expect.type<Boom.Boom>(Boom.clientTimeout('timed out', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.clientTimeout('timed out'));
expect.type<Boom.Boom>(Boom.clientTimeout());

expect.error(Boom.clientTimeout(408));
expect.error(Boom.clientTimeout({ foo: 'bar' }));


// conflict()

expect.type<Boom.Boom>(Boom.conflict('there was a conflict', 'some data'));
expect.type<Boom.Boom>(Boom.conflict('there was a conflict', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.conflict('there was a conflict'));
expect.type<Boom.Boom>(Boom.conflict());

expect.error(Boom.conflict(409));
expect.error(Boom.conflict({ foo: 'bar' }));


// resourceGone()

expect.type<Boom.Boom>(Boom.resourceGone('it is gone', 'some data'));
expect.type<Boom.Boom>(Boom.resourceGone('it is gone', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.resourceGone('it is gone'));
expect.type<Boom.Boom>(Boom.resourceGone());

expect.error(Boom.resourceGone(410));
expect.error(Boom.resourceGone({ foo: 'bar' }));


// lengthRequired()

expect.type<Boom.Boom>(Boom.lengthRequired('length needed', 'some data'));
expect.type<Boom.Boom>(Boom.lengthRequired('length needed', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.lengthRequired('length needed'));
expect.type<Boom.Boom>(Boom.lengthRequired());

expect.error(Boom.lengthRequired(411));
expect.error(Boom.lengthRequired({ foo: 'bar' }));


// preconditionFailed()

expect.type<Boom.Boom>(Boom.preconditionFailed('failed', 'some data'));
expect.type<Boom.Boom>(Boom.preconditionFailed('failed', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.preconditionFailed('failed'));
expect.type<Boom.Boom>(Boom.preconditionFailed());

expect.error(Boom.preconditionFailed(412));
expect.error(Boom.preconditionFailed({ foo: 'bar' }));


// entityTooLarge()

expect.type<Boom.Boom>(Boom.entityTooLarge('too big', 'some data'));
expect.type<Boom.Boom>(Boom.entityTooLarge('too big', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.entityTooLarge('too big'));
expect.type<Boom.Boom>(Boom.entityTooLarge());

expect.error(Boom.entityTooLarge(413));
expect.error(Boom.entityTooLarge({ foo: 'bar' }));


// uriTooLong()

expect.type<Boom.Boom>(Boom.uriTooLong('uri is too long', 'some data'));
expect.type<Boom.Boom>(Boom.uriTooLong('uri is too long', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.uriTooLong('uri is too long'));
expect.type<Boom.Boom>(Boom.uriTooLong());

expect.error(Boom.uriTooLong(414));
expect.error(Boom.uriTooLong({ foo: 'bar' }));


// unsupportedMediaType()

expect.type<Boom.Boom>(Boom.unsupportedMediaType('that media is not supported', 'some data'));
expect.type<Boom.Boom>(Boom.unsupportedMediaType('that media is not supported', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.unsupportedMediaType('that media is not supported'));
expect.type<Boom.Boom>(Boom.unsupportedMediaType());

expect.error(Boom.unsupportedMediaType(415));
expect.error(Boom.unsupportedMediaType({ foo: 'bar' }));


// rangeNotSatisfiable()

expect.type<Boom.Boom>(Boom.rangeNotSatisfiable('range not satisfiable', 'some data'));
expect.type<Boom.Boom>(Boom.rangeNotSatisfiable('range not satisfiable', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.rangeNotSatisfiable('range not satisfiable'));
expect.type<Boom.Boom>(Boom.rangeNotSatisfiable());

expect.error(Boom.rangeNotSatisfiable(416));
expect.error(Boom.rangeNotSatisfiable({ foo: 'bar' }));


// expectationFailed()

expect.type<Boom.Boom>(Boom.expectationFailed('expected this to work', 'some data'));
expect.type<Boom.Boom>(Boom.expectationFailed('expected this to work', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.expectationFailed('expected this to work'));
expect.type<Boom.Boom>(Boom.expectationFailed());

expect.error(Boom.expectationFailed(417));
expect.error(Boom.expectationFailed({ foo: 'bar' }));


// teapot()

expect.type<Boom.Boom>(Boom.teapot('sorry, no coffee...', 'some data'));
expect.type<Boom.Boom>(Boom.teapot('sorry, no coffee...', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.teapot('sorry, no coffee...'));
expect.type<Boom.Boom>(Boom.teapot());

expect.error(Boom.teapot(418));
expect.error(Boom.teapot({ foo: 'bar' }));


// badData()

expect.type<Boom.Boom>(Boom.badData('your data is bad and you should feel bad', 'some data'));
expect.type<Boom.Boom>(Boom.badData('your data is bad and you should feel bad', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.badData('your data is bad and you should feel bad'));
expect.type<Boom.Boom>(Boom.badData());

expect.error(Boom.badData(422));
expect.error(Boom.badData({ foo: 'bar' }));


// locked()

expect.type<Boom.Boom>(Boom.locked('this resource has been locked', 'some data'));
expect.type<Boom.Boom>(Boom.locked('this resource has been locked', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.locked('this resource has been locked'));
expect.type<Boom.Boom>(Boom.locked());

expect.error(Boom.locked(423));
expect.error(Boom.locked({ foo: 'bar' }));


// failedDependency()

expect.type<Boom.Boom>(Boom.failedDependency('an external resource failed', 'some data'));
expect.type<Boom.Boom>(Boom.failedDependency('an external resource failed', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.failedDependency('an external resource failed'));
expect.type<Boom.Boom>(Boom.failedDependency());

expect.error(Boom.failedDependency(424));
expect.error(Boom.failedDependency({ foo: 'bar' }));


// preconditionRequired()

expect.type<Boom.Boom>(Boom.preconditionRequired('you must supple an If-Match header', 'some data'));
expect.type<Boom.Boom>(Boom.preconditionRequired('you must supple an If-Match header', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.preconditionRequired('you must supple an If-Match header'));
expect.type<Boom.Boom>(Boom.preconditionRequired());

expect.error(Boom.preconditionRequired(428));
expect.error(Boom.preconditionRequired({ foo: 'bar' }));


// tooManyRequests()

expect.type<Boom.Boom>(Boom.tooManyRequests('you have exceeded your request limit', 'some data'));
expect.type<Boom.Boom>(Boom.tooManyRequests('you have exceeded your request limit', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.tooManyRequests('you have exceeded your request limit'));
expect.type<Boom.Boom>(Boom.tooManyRequests());

expect.error(Boom.tooManyRequests(414));
expect.error(Boom.tooManyRequests({ foo: 'bar' }));


// illegal()

expect.type<Boom.Boom>(Boom.illegal('you are not permitted to view this resource for legal reasons', 'some data'));
expect.type<Boom.Boom>(Boom.illegal('you are not permitted to view this resource for legal reasons', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.illegal('you are not permitted to view this resource for legal reasons'));
expect.type<Boom.Boom>(Boom.illegal());

expect.error(Boom.illegal(451));
expect.error(Boom.illegal({ foo: 'bar' }));


// 5xx Errors

// internal()

expect.type<Boom.Boom>(Boom.internal('terrible implementation', 'some data', 599));
expect.type<Boom.Boom>(Boom.internal('terrible implementation', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.internal('terrible implementation'));
expect.type<Boom.Boom>(Boom.internal());

expect.error(Boom.internal(500));
expect.error(Boom.internal({ foo: 'bar' }));


// badImplementation()

expect.type<Boom.Boom>(Boom.badImplementation('terrible implementation', 'some data'));
expect.type<Boom.Boom>(Boom.badImplementation('terrible implementation', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.badImplementation('terrible implementation'));
expect.type<Boom.Boom>(Boom.badImplementation());

expect.error(Boom.badImplementation(500));
expect.error(Boom.badImplementation({ foo: 'bar' }));


// notImplemented()

expect.type<Boom.Boom>(Boom.notImplemented('method not implemented', 'some data'));
expect.type<Boom.Boom>(Boom.notImplemented('method not implemented', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.notImplemented('method not implemented'));
expect.type<Boom.Boom>(Boom.notImplemented());

expect.error(Boom.notImplemented(501));
expect.error(Boom.notImplemented({ foo: 'bar' }));


// badGateway()

expect.type<Boom.Boom>(Boom.badGateway('this is a bad gateway', 'some data'));
expect.type<Boom.Boom>(Boom.badGateway('this is a bad gateway', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.badGateway('this is a bad gateway'));
expect.type<Boom.Boom>(Boom.badGateway());

expect.error(Boom.badGateway(502));
expect.error(Boom.badGateway({ foo: 'bar' }));


// serverUnavailable()

expect.type<Boom.Boom>(Boom.serverUnavailable('unavailable', 'some data'));
expect.type<Boom.Boom>(Boom.serverUnavailable('unavailable', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.serverUnavailable('unavailable'));
expect.type<Boom.Boom>(Boom.serverUnavailable());

expect.error(Boom.serverUnavailable(503));
expect.error(Boom.serverUnavailable({ foo: 'bar' }));


// gatewayTimeout()

expect.type<Boom.Boom>(Boom.gatewayTimeout('gateway timeout', 'some data'));
expect.type<Boom.Boom>(Boom.gatewayTimeout('gateway timeout', { foo: 'bar' }));
expect.type<Boom.Boom>(Boom.gatewayTimeout('gateway timeout'));
expect.type<Boom.Boom>(Boom.gatewayTimeout());

expect.error(Boom.gatewayTimeout(504));
expect.error(Boom.gatewayTimeout({ foo: 'bar' }));


lib/http-browser.js
var STATUS_CODES = {
  '100': 'Continue',
  '101': 'Switching Protocols',
  '102': 'Processing',
  '200': 'OK',
  '201': 'Created',
  '202': 'Accepted',
  '203': 'Non-Authoritative Information',
  '204': 'No Content',
  '205': 'Reset Content',
  '206': 'Partial Content',
  '207': 'Multi-Status',
  '300': 'Multiple Choices',
  '301': 'Moved Permanently',
  '302': 'Moved Temporarily',
  '303': 'See Other',
  '304': 'Not Modified',
  '305': 'Use Proxy',
  '307': 'Temporary Redirect',
  '308': 'Permanent Redirect',
  '400': 'Bad Request',
  '401': 'Unauthorized',
  '402': 'Payment Required',
  '403': 'Forbidden',
  '404': 'Not Found',
  '405': 'Method Not Allowed',
  '406': 'Not Acceptable',
  '407': 'Proxy Authentication Required',
  '408': 'Request Time-out',
  '409': 'Conflict',
  '410': 'Gone',
  '411': 'Length Required',
  '412': 'Precondition Failed',
  '413': 'Request Entity Too Large',
  '414': 'Request-URI Too Large',
  '415': 'Unsupported Media Type',
  '416': 'Requested Range Not Satisfiable',
  '417': 'Expectation Failed',
  '418': 'I\'m a teapot',
  '422': 'Unprocessable Entity',
  '423': 'Locked',
  '424': 'Failed Dependency',
  '425': 'Unordered Collection',
  '426': 'Upgrade Required',
  '428': 'Precondition Required',
  '429': 'Too Many Requests',
  '431': 'Request Header Fields Too Large',
  '500': 'Internal Server Error',
  '501': 'Not Implemented',
  '502': 'Bad Gateway',
  '503': 'Service Unavailable',
  '504': 'Gateway Time-out',
  '505': 'HTTP Version Not Supported',
  '506': 'Variant Also Negotiates',
  '507': 'Insufficient Storage',
  '509': 'Bandwidth Limit Exceeded',
  '510': 'Not Extended',
  '511': 'Network Authentication Required'
};

module.exports = {
  STATUS_CODES: STATUS_CODES
};

lib/index.js
'use strict';

const Hoek = require('@hapi/hoek');


const internals = {
    codes: new Map([
        [100, 'Continue'],
        [101, 'Switching Protocols'],
        [102, 'Processing'],
        [200, 'OK'],
        [201, 'Created'],
        [202, 'Accepted'],
        [203, 'Non-Authoritative Information'],
        [204, 'No Content'],
        [205, 'Reset Content'],
        [206, 'Partial Content'],
        [207, 'Multi-Status'],
        [300, 'Multiple Choices'],
        [301, 'Moved Permanently'],
        [302, 'Moved Temporarily'],
        [303, 'See Other'],
        [304, 'Not Modified'],
        [305, 'Use Proxy'],
        [307, 'Temporary Redirect'],
        [400, 'Bad Request'],
        [401, 'Unauthorized'],
        [402, 'Payment Required'],
        [403, 'Forbidden'],
        [404, 'Not Found'],
        [405, 'Method Not Allowed'],
        [406, 'Not Acceptable'],
        [407, 'Proxy Authentication Required'],
        [408, 'Request Time-out'],
        [409, 'Conflict'],
        [410, 'Gone'],
        [411, 'Length Required'],
        [412, 'Precondition Failed'],
        [413, 'Request Entity Too Large'],
        [414, 'Request-URI Too Large'],
        [415, 'Unsupported Media Type'],
        [416, 'Requested Range Not Satisfiable'],
        [417, 'Expectation Failed'],
        [418, 'I\'m a teapot'],
        [422, 'Unprocessable Entity'],
        [423, 'Locked'],
        [424, 'Failed Dependency'],
        [425, 'Unordered Collection'],
        [426, 'Upgrade Required'],
        [428, 'Precondition Required'],
        [429, 'Too Many Requests'],
        [431, 'Request Header Fields Too Large'],
        [451, 'Unavailable For Legal Reasons'],
        [500, 'Internal Server Error'],
        [501, 'Not Implemented'],
        [502, 'Bad Gateway'],
        [503, 'Service Unavailable'],
        [504, 'Gateway Time-out'],
        [505, 'HTTP Version Not Supported'],
        [506, 'Variant Also Negotiates'],
        [507, 'Insufficient Storage'],
        [509, 'Bandwidth Limit Exceeded'],
        [510, 'Not Extended'],
        [511, 'Network Authentication Required']
    ])
};


exports.Boom = class extends Error {

    constructor(message, options = {}) {

        if (message instanceof Error) {
            return exports.boomify(Hoek.clone(message), options);
        }

        const { statusCode = 500, data = null, ctor = exports.Boom } = options;
        const error = new Error(message ? message : undefined);         // Avoids settings null message
        Error.captureStackTrace(error, ctor);                           // Filter the stack to our external API
        error.data = data;
        const boom = internals.initialize(error, statusCode);

        Object.defineProperty(boom, 'typeof', { value: ctor });

        if (options.decorate) {
            Object.assign(boom, options.decorate);
        }

        return boom;
    }

    static [Symbol.hasInstance](instance) {

        return exports.isBoom(instance);
    }
};


exports.isBoom = function (err) {

    return err instanceof Error && !!err.isBoom;
};


exports.boomify = function (err, options) {

    Hoek.assert(err instanceof Error, 'Cannot wrap non-Error object');

    options = options || {};

    if (options.data !== undefined) {
        err.data = options.data;
    }

    if (options.decorate) {
        Object.assign(err, options.decorate);
    }

    if (!err.isBoom) {
        return internals.initialize(err, options.statusCode || 500, options.message);
    }

    if (options.override === false ||                           // Defaults to true
        !options.statusCode && !options.message) {

        return err;
    }

    return internals.initialize(err, options.statusCode || err.output.statusCode, options.message);
};


// 4xx Client Errors

exports.badRequest = function (message, data) {

    return new exports.Boom(message, { statusCode: 400, data, ctor: exports.badRequest });
};


exports.unauthorized = function (message, scheme, attributes) {          // Or (message, wwwAuthenticate[])

    const err = new exports.Boom(message, { statusCode: 401, ctor: exports.unauthorized });

    // function (message)

    if (!scheme) {
        return err;
    }

    // function (message, wwwAuthenticate[])

    if (typeof scheme !== 'string') {
        err.output.headers['WWW-Authenticate'] = scheme.join(', ');
        return err;
    }

    // function (message, scheme, attributes)

    let wwwAuthenticate = `${scheme} `;

    if (attributes ||
        message) {

        err.output.payload.attributes = {};
    }

    if (attributes) {
        if (typeof attributes === 'string') {
            wwwAuthenticate += Hoek.escapeHeaderAttribute(attributes);
            err.output.payload.attributes = attributes;
        }
        else {
            wwwAuthenticate += Object.keys(attributes).map((name) => {

                let value = attributes[name];
                if (value === null ||
                    value === undefined) {

                    value = '';
                }

                err.output.payload.attributes[name] = value;
                return `${name}="${Hoek.escapeHeaderAttribute(value.toString())}"`;
            })
                .join(', ');
        }
    }

    if (message) {
        if (attributes) {
            wwwAuthenticate += ', ';
        }

        wwwAuthenticate += `error="${Hoek.escapeHeaderAttribute(message)}"`;
        err.output.payload.attributes.error = message;
    }
    else {
        err.isMissing = true;
    }

    err.output.headers['WWW-Authenticate'] = wwwAuthenticate;
    return err;
};


exports.paymentRequired = function (message, data) {

    return new exports.Boom(message, { statusCode: 402, data, ctor: exports.paymentRequired });
};


exports.forbidden = function (message, data) {

    return new exports.Boom(message, { statusCode: 403, data, ctor: exports.forbidden });
};


exports.notFound = function (message, data) {

    return new exports.Boom(message, { statusCode: 404, data, ctor: exports.notFound });
};


exports.methodNotAllowed = function (message, data, allow) {

    const err = new exports.Boom(message, { statusCode: 405, data, ctor: exports.methodNotAllowed });

    if (typeof allow === 'string') {
        allow = [allow];
    }

    if (Array.isArray(allow)) {
        err.output.headers.Allow = allow.join(', ');
    }

    return err;
};


exports.notAcceptable = function (message, data) {

    return new exports.Boom(message, { statusCode: 406, data, ctor: exports.notAcceptable });
};


exports.proxyAuthRequired = function (message, data) {

    return new exports.Boom(message, { statusCode: 407, data, ctor: exports.proxyAuthRequired });
};


exports.clientTimeout = function (message, data) {

    return new exports.Boom(message, { statusCode: 408, data, ctor: exports.clientTimeout });
};


exports.conflict = function (message, data) {

    return new exports.Boom(message, { statusCode: 409, data, ctor: exports.conflict });
};


exports.resourceGone = function (message, data) {

    return new exports.Boom(message, { statusCode: 410, data, ctor: exports.resourceGone });
};


exports.lengthRequired = function (message, data) {

    return new exports.Boom(message, { statusCode: 411, data, ctor: exports.lengthRequired });
};


exports.preconditionFailed = function (message, data) {

    return new exports.Boom(message, { statusCode: 412, data, ctor: exports.preconditionFailed });
};


exports.entityTooLarge = function (message, data) {

    return new exports.Boom(message, { statusCode: 413, data, ctor: exports.entityTooLarge });
};


exports.uriTooLong = function (message, data) {

    return new exports.Boom(message, { statusCode: 414, data, ctor: exports.uriTooLong });
};


exports.unsupportedMediaType = function (message, data) {

    return new exports.Boom(message, { statusCode: 415, data, ctor: exports.unsupportedMediaType });
};


exports.rangeNotSatisfiable = function (message, data) {

    return new exports.Boom(message, { statusCode: 416, data, ctor: exports.rangeNotSatisfiable });
};


exports.expectationFailed = function (message, data) {

    return new exports.Boom(message, { statusCode: 417, data, ctor: exports.expectationFailed });
};


exports.teapot = function (message, data) {

    return new exports.Boom(message, { statusCode: 418, data, ctor: exports.teapot });
};


exports.badData = function (message, data) {

    return new exports.Boom(message, { statusCode: 422, data, ctor: exports.badData });
};


exports.locked = function (message, data) {

    return new exports.Boom(message, { statusCode: 423, data, ctor: exports.locked });
};


exports.failedDependency = function (message, data) {

    return new exports.Boom(message, { statusCode: 424, data, ctor: exports.failedDependency });
};


exports.preconditionRequired = function (message, data) {

    return new exports.Boom(message, { statusCode: 428, data, ctor: exports.preconditionRequired });
};


exports.tooManyRequests = function (message, data) {

    return new exports.Boom(message, { statusCode: 429, data, ctor: exports.tooManyRequests });
};


exports.illegal = function (message, data) {

    return new exports.Boom(message, { statusCode: 451, data, ctor: exports.illegal });
};


// 5xx Server Errors

exports.internal = function (message, data, statusCode = 500) {

    return internals.serverError(message, data, statusCode, exports.internal);
};


exports.notImplemented = function (message, data) {

    return internals.serverError(message, data, 501, exports.notImplemented);
};


exports.badGateway = function (message, data) {

    return internals.serverError(message, data, 502, exports.badGateway);
};


exports.serverUnavailable = function (message, data) {

    return internals.serverError(message, data, 503, exports.serverUnavailable);
};


exports.gatewayTimeout = function (message, data) {

    return internals.serverError(message, data, 504, exports.gatewayTimeout);
};


exports.badImplementation = function (message, data) {

    const err = internals.serverError(message, data, 500, exports.badImplementation);
    err.isDeveloperError = true;
    return err;
};


internals.initialize = function (err, statusCode, message) {

    const numberCode = parseInt(statusCode, 10);
    Hoek.assert(!isNaN(numberCode) && numberCode >= 400, 'First argument must be a number (400+):', statusCode);

    err.isBoom = true;
    err.isServer = numberCode >= 500;

    if (!err.hasOwnProperty('data')) {
        err.data = null;
    }

    err.output = {
        statusCode: numberCode,
        payload: {},
        headers: {}
    };

    Object.defineProperty(err, 'reformat', { value: internals.reformat });

    if (!message &&
        !err.message) {

        err.reformat();
        message = err.output.payload.error;
    }

    if (message) {
        const props = Object.getOwnPropertyDescriptor(err, 'message') || Object.getOwnPropertyDescriptor(Object.getPrototypeOf(err), 'message');
        Hoek.assert(!props || props.configurable && !props.get, 'The error is not compatible with boom');

        err.message = message + (err.message ? ': ' + err.message : '');
        err.output.payload.message = err.message;
    }

    err.reformat();
    return err;
};


internals.reformat = function (debug = false) {

    this.output.payload.statusCode = this.output.statusCode;
    this.output.payload.error = internals.codes.get(this.output.statusCode) || 'Unknown';

    if (this.output.statusCode === 500 && debug !== true) {
        this.output.payload.message = 'An internal server error occurred';              // Hide actual error from user
    }
    else if (this.message) {
        this.output.payload.message = this.message;
    }
};


internals.serverError = function (message, data, statusCode, ctor) {

    if (data instanceof Error &&
        !data.isBoom) {

        return exports.boomify(data, { statusCode, message });
    }

    return new exports.Boom(message, { statusCode, data, ctor });
};
